### ISSUE TITLE
Issue with delta.time.secs

### ISSUE THREAD

**Issue by kaitlyngaynor** [wildcam-gorongosa/data-cleaning#5]
From e-mails: 

MEREDITH:

Here's the issue I ran into with generating independent (>=15 minute) records with the WildCam data. This may just be an issue with my version of R (or of the camtrapR package or use of strptime vs. as.Date or whatever...), but worth taking a look just in case?    

This is the parsing code from your shiny app: 

records_subset <- reactive({
    records %>%
      filter(delta.time.secs == 0 | delta.time.secs >= (60 * input$independent_min),
             Date >= input$date_range[1], Date <= input$date_range[2]) %>%
      filter(Species == input$species_select)
  })
  
This code looks at the delta.time.secs value and only retains records where delta.time.secs == 0 (the first independent sighting) or is in a sequence of sightings but greater than/equal to the independent_min # of minutes apart. 

I don't believe you calculate the delta.time.secs in your Shiny app code but I'm assuming you used the 'diff()' function? I had to regenerate the delta.time for the datasheet set-up that I was working with and found an issue where this function often rounded the number of seconds down to 0 even when the number of seconds WAS greater than 0 between consecutive records. Everything would look like it ran fine, but if you actually dug deep into the resulting dataframe, a lot of "0"s would be retained that were actually non-independent records taken within 15 minutes of each other. 

This issue went away when I calculated the time difference in minutes instead of seconds, i.e., this -- 

fulldat$delta.time.mins <- unlist(tapply(fulldat$datetime, INDEX = fulldat$index,
                                     FUN = function(x) c(0, `units<-`(diff(x), "mins"))))
fulldat <- fulldat %>% filter(delta.time.mins == 0 | delta.time.mins >= 15) 
 
-- produces the dataset I was hoping for. 

??? R ??? 

Hopefully a non-issue for you but FYI!

**Comment by kaitlyngaynor** [comment]
KAITLYN: 

Ooh good to know. I had a script in the Shiny app that did this on the non-camtrapR data as follows, in a very ugly for loop. I actually used "difftime" instead of "diff" and am not sure if that makes a difference? I can have another look and see if this still raises issues...

records$delta.time.secs <- NA # create column for delta time
records$delta.time.secs[[1]] <- 0 # set first row to 0

for (i in 2:nrow(records)) {
  
  if(records$Camera[[i]] == records$Camera[[i-1]] &
     records$species[[i]] == records$species[[i-1]]) {
    
    # calculate difference 
    records$delta.time.secs[[i]] <- difftime(records$datetime[[i]], 
                                             records$datetime[[i-1]], 
                                             units = "secs")

  } else {
    
    records$delta.time.secs[[i]] <- 0
    
  }
  
}

**Comment by kaitlyngaynor** [comment]
KAITLYN: 

One more thing: I also sorted by camera, then species, then date-time before applying the function
records <- records[order(records$Camera, records$species, records$datetime),]
which is pretty important. assuming you did this too somewhere, though?

I figured I'd try to run your script and mine, and then see where the filtering broke down and resulted in different dataframes (if at all—hopefully not at all!). But when I run the script that you sent, I end up with 0s for the entire column of delta.time.mins. What is in your fulldat$index column? I don't have that in the dataframe I'm using, so I just created an integer sequence / row number and used that... not sure if that's the issue or what.

Anyway, if you have the full script + input data I can try to reproduce on my end and get to the bottom of this! hopefully a nonissue but I want to be sure...

**Comment by kaitlyngaynor** [comment]
KAITLYN:

Good news, your method and my method result in identical cleaned datasets! With one small addition needed—the 2019 data were rounded to the nearest minute, whereas the previous data were not, which meant that it was necessary to run these lines that you added:

fulldat$tag <- paste(fulldat$site, fulldat$species, fulldat$datetime)
fulldat <- fulldat[!duplicated(fulldat$tag),]

I'm not sure if that was the issue you were running into, but using seconds rather than minutes still works fine for me, so long as I run those two lines first. Good to know for prepping data for the Shiny app, though—I had to add those two lines in before doing the delta.time.secs bit.

I added a script to our data-cleaning repo just as a record!

Your way runs a lot faster btw :D
